### PR TITLE
Add Tutorial for how to connect a Docker Compose workflow to Kubernetes

### PR DESCRIPTION
**Pull request by ezodude**
Resolves https://github.com/appvia/kev/issues/203

Assumes the theme of connecting a Docker Compose workflow to Kubernetes.

It's based on the problems developers encounter when migrating to Kubernetes. This was researched [using these articles](https://www.google.com/search?hl=en&q=migrating%20a%20docker%20compose%20workflow%20to%20kubernetes).

Contains,
- A new `Installation` section.
- A refactored `Quickstart` section.
- A new `Tutorial` section.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
 
-### Initiate a project
+## Quickstart
 
-Run the following command from within your project directory to initialise Kev project:
+- `kev init` - identifies a project's Compose Kubernetes source files and creates Compose environment overrides.
+- `kev render` - detects, applies any config changes and generates deployment manifests.
+- `kev help` - run it if you're a little lost.
+
+### init
```
Comment (marcinc) [inline README.md:44]: Perhaps change title of this section to be more descriptive.. e.g. `Project initialisation`, `Initialise project` or similar?

File: README.md [hunk 2]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+
+Run the following command within your project directory:
 
 ```sh
 kev init
 ```
 
-The command above will auto-detect default `docker-compose.yaml` and `docker-compose.override.yaml` files (if present) in the project directory, and tracks them as Kubernetes deployment sources.
+This identifies the default `docker-compose.yaml` and (if present) `docker-compose.override.yaml` files in your project directory. They will be used as Compose Kubernetes sources.
```
Comment (marcinc) [inline README.md:52]: need to clarify that so it's less confusing... `as Compose Kubernetes sources` -> perhaps `as source of truth for your application deployment in Kubernetes`?

File: README.md [hunk 3]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
 
-If you want to point _Kev_ at set of alternative compose files, simply pass them in with `-f` flag. Multiple source compose files can be specified by providing `-f` flag multiple times.
+Also, it creates a default `dev` environment and its Compose override file.  
 
-And, to individually control the configuration of your project for deployments in separate target environments, just specify their respective names with `-e` flag, e.g.:
+Here's another example. It uses an alternate `docker-compose` file with `stage` + `prod` environments:
```
Comment (marcinc) [inline README.md:56]: `+` -> `&`

File: README.md [hunk 4]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
-├── docker-compose.kev.prod.yaml  # prod environment override
-├── kev.yaml                      # kev project manifest
+├── docker-compose.kev.stage.yaml       # stage Compose environment override file
+├── docker-compose.kev.prod.yaml        # prod Compose environment override file
+├── kev.yaml                            # kev project manifest
 ├── ...
 ```
 
-Your project is now ready and environment specific configuration files generated. _Kev_ will infer useful configuration details already present in the source compose file(s). If this is not possible it'll assume sensible defaults. Users have an easy way to manually alter values directly in the environment specific override file as and when required.
+Here's what happened, _Kev_,
```
Comment (marcinc) [inline README.md:75]: Rephrase: 

```
Here's what happened... _Kev_ has:
- Inferred the configuration details already present in your docker compose source files.
...
```

File: README.md [hunk 5]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+- Inferred the configuration details already present in your compose Kubernetes deployment sources.
+- Assigned sensible defaults for any config it couldn't infer.
+- Created Compose overrides files for the `stage` and `prod` environments.
+
+That's it, _Kev_ is now bootstrapped and ready!
```
Comment (marcinc) [inline README.md:80]: `That's it, _Kev_ is now bootstrapped and ready!` -> `That's it, your _Kev_ project is now bootstrapped and ready!`

File: README.md [hunk 6]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+From now on it can,
+- Detect edits in your source compose file.
+- Apply any related config changes to your compose environment overrides.
+- Generate deployment manifests.   
 
-In order to deploy your project application to Kubernetes you will need to supply _something_ K8s can understand. Currently, _Kev_ only supports Docker Compose conversion to native Kuberentes manifests. The Community might add additional output formats at later stages. See our [Roadmap][roadmap] for planned new features.
+You can now customise your deployment targets by altering values in the relevant Compose environment override file. 
 
-To render manifests, simply run the following command from your project root:
+### render
```
Comment (marcinc) [inline README.md:89]: `render` -> `Render K8s manifests`

File: README.md [hunk 7]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+We now need to generate manifests based on your Docker Compose config and environments. You'll use these manifests to deploy your app to Kubernetes. 
+
+Run the following command from your project root:
 
 ```sh
 kev render
 ```
 
-`render` generates the project's manifests based on the tracked docker-compose files, using the desired output format specified via the `-f` flag (defaults to `kubernetes`), and for selected environments. Note that ALL environments will be rendered by default if none are specified.
+The command,
```
Comment (marcinc) [inline README.md:99]: `The command,` -> `The command above,`

File: README.md [hunk 8]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+- Detects edits you made to the project's source compose file(s) to re-infer config changes.
```
Comment (marcinc) [inline README.md:100]: drop ` to re-infer config changes` as it's covered by the second item already.

File: README.md [hunk 9]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+- Applies any found config changes to your compose environment overrides.
+- Generates kubernetes manifests based on all compose files including environment overrides.
+- Targets all environments.
```
Comment (marcinc) [inline README.md:103]: rephrase `Targets all environments` -> `Produces manifests for all environments`

File: README.md [hunk 10]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
 
-You can control which environments to render the manifests for with `-e` flag(s).
+The directory below should now appear in your project directory:
 
-You may also specify the output directory for generated manifests with `-d` flag. Note, that specified directory will contain sub-directories, each for a separate environment name for which manifests were generated.
+```sh
+├── k8s     # stores the Kubernetes manifests for all target deployment environments. 
+├── ...
```
Comment (marcinc) [inline README.md:109]: ^ would be good to give examples including subdirectories for stages from the example above.

File: README.md [hunk 11]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
 
-Note that all generated manifests are fully expanded i.e. they should not be treated as templates. (Quick reminder that a specific environment configuration lives in the docker compose override files.)
+**Note:** Generated manifests should **NOT** be treated as templates as they are fully expanded.
 
-From this point onward you're free to use whatever tool or framework you are already familiar with, to deploy your project to Kubernetes e.g `kubectl`, `skaffold` etc. Watch our [Roadmap][roadmap] for details around planned integrations.
+### How can I deploy the app to Kubernetes?
 
-## Commands
+To deploy your app onto Kubernetes,
+- Ensure you can access a running Kubernetes installation, either locally or remotely.
```
Comment (marcinc) [inline README.md:123]: Link `locally` to something like minikube or docker-for-desktop etc?

File: README.md [hunk 12]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+- Use `kubectl` to apply the manifests.
```
Comment (marcinc) [inline README.md:124]: link `kubectl` to official page with download / installation instructions etc.

File: README.md [hunk 13]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
 
-- `kev init` - initiate project for kev.
-- `kev render` - render application manifests for selected environments, according to desired output format.
-- `kev help` - run it if you're a little lost.
+In this example, we deploy the stage environment:
```
Comment (marcinc) [inline README.md:126]: stage -> `stage`

**Review by marcinc — COMMENTED** [review]
initial comments to the quick start section. moving onto Tutorial.

File: README.md (hunk 1 above)
Comment (ezodude) [reply README.md:44]: @marcinc It's meant to mirror the command `init`.

I played with the other titles, but thought the simpler title would be ideal for a `Quickstart`.

File: README.md (hunk 6 above)
Comment (ezodude) [reply README.md:89]: Similar rationale to `init` as explained above.

File: README.md (hunk 10 above)
Comment (ezodude) [reply README.md]: done

**Review by marcinc — CHANGES_REQUESTED** [review]
A bunch of change suggestions and comments. Let's extract tutorial to a separate .md file and link it in the README. Similar approach for other additional follow-up write-ups.

File: README.md [hunk 14]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+$ docker-compose down -v    # Stop all containers. Remove named volumes.
+```
+
+### Preparing for Kubernetes
+
+Compose and Kubernetes address different problems.
+
+_Compose_, helps you wire and run your application on a single machine with a single instance running for every service. It's super for development.
+
+_Kubernetes_, however, is designed to help you run your application in a highly available mode using clusters and service replications. It's build for production like environments.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:207]: I think ... `It is production grade` will suffice.

File: README.md [hunk 15]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+
+Describing Compose services as Kubernetes services requires an extra layer that translates between these different worlds.
+
+Furthermore, since you're moving your app to Kubernetes, you're moving from a development mindset to sharing your app for testing, staging and at some point publishing to a production environment.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:211]: I see what you're saying will reword to capture the gist.

File: README.md [hunk 16]
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+```shell script
+$ kev init -e local -e stage
+# ✓ Init
+#  → Creating kev.yaml ... Done
+#  → Creating docker-compose.kev.local.yaml ... Done
+#  → Creating docker-compose.kev.stage.yaml ... Done
+```
+
+_Kev_ has now been initialised and configured. It has,
+- Started tracking the `docker-compose.yaml` file as a Compose Kubernetes source.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:236]: I'm trying to establish a vocabulary to explain key concepts, hence `Compose Kubernetes source`.

I'll have a think to see it still makes sense.

File: README.md [hunk 17]
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+
+We'll be adjusting these values soon per target environment. For now, they look as below,
+
+```yaml
+version: "3.7"
+services:
+  wordpress:
+    labels:
+      kev.service.type: LoadBalancer
+      kev.workload.cpu: "0.1"
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:270]: I agree ... it really bugs me.

File: README.md (hunk 1 above)
Comment (ezodude) [reply README.md:44]: embed into render.

File: README.md [hunk 18]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+
+## Tutorial
+
+This tutorial will walk you through how to **connect your Docker Compose Workflow to Kubernetes** - using _Kev_.
+
+This is NOT a migration.
+
+Your hard-earned Docker Compose skills will help you work with Kubernetes speeding up development and iteration.   
+
+We'll set up _Kev_, iterate and deploy a [WordPress application]() onto Kubernetes.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:149]: Not that one, it's has the wrong image. Now linking to https://docs.docker.com/compose/wordpress/

File: README.md [hunk 19]
```diff
@@ -31,72 +31,582 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+Ace, we're all good, let's stop the service,
+```shell script
+$ docker-compose down -v    # Stop all containers. Remove named volumes.
+```
+
+### Preparing for Kubernetes
+
+Compose and Kubernetes address different problems.
+
+_Compose_, helps you wire and run your application on a single machine with a single instance running for every service. It's super for development.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:205]: I get the gist, re-phrasing to make sense in the greater context.

File: README.md [hunk 20]
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+# service/db created
+# statefulset.apps/db created
+# networkpolicy.networking.k8s.io/default configured
+# deployment.apps/wordpress configured
+# service/wordpress configured
+```
+
+**This BREAKS our running app** - to fix it we need to understand how service find-ability differs between Compose and Kubernetes.
+
+#### Make db service findable in Kubernetes
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:495]: Going for: `Fix db service discovery`

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:236]: Let's got with: `as the source application manifest`

... so should read:
- Started tracking the `docker-compose.yaml` file as the source application manifest.

File: README.md [hunk 21]
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+
+_Kev_ has now been initialised and configured. It has,
+- Started tracking the `docker-compose.yaml` file as a Compose Kubernetes source.
+- Inferred configuration details from the `docker-compose.yaml` file.
+- Assigned sensible defaults for any config it couldn't infer.
+- Created `local` (useful for a testing our own machine) and `staging` (useful for a testing our a remote machine) _Compose environment overrides_.
+
+It has also generated three files:
+- `kev.yaml`, a manifest that describes our _Compose Kubernetes sources_.
+- `docker-compose.kev.*.yaml`, two files to represent our _Compose environment overrides_.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:243]: Will leave it 🙂

File: README.md [hunk 22]
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+# INFO 💡: 🖨️  Rendering stage environment
+# INFO 💡: Target Dir: k8s/stage
+# INFO 💡: ⎈  kubernetes file "k8s/stage/wordpress-service.yaml" created
+# INFO 💡: ⎈  kubernetes file "k8s/stage/wordpress-deployment.yaml" created
+# INFO 💡: ⎈  kubernetes file "k8s/stage/default-networkpolicy.yaml" created
+# INFO 💡: 🧰 App render complete!
+```
+
+In this case, _Kev_,
+- Has re-introspected our _Compose Kubernetes source_.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:324]: `_Source application manifest_`

File: README.md [hunk 23]
```diff
@@ -31,72 +31,585 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+```
+
+_Kev_ has now been initialised and configured. It has,
+- Started tracking the `docker-compose.yaml` file as a Compose Kubernetes source.
+- Inferred configuration details from the `docker-compose.yaml` file.
+- Assigned sensible defaults for any config it couldn't infer.
+- Created `local` (useful for a testing our own machine) and `staging` (useful for a testing our a remote machine) _Compose environment overrides_.
+
+It has also generated three files:
+- `kev.yaml`, a manifest that describes our _Compose Kubernetes sources_.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:242]: reworded to ... `our _Source application manifest_ and _Compose environment overrides_`

**Review by marcinc — COMMENTED** [review]
A few more comments

File: README.md [hunk 24]
```diff
@@ -31,72 +31,586 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
+
+Compose and Kubernetes address different problems.
+
+_Compose_, helps you wire, develop and run your application locally as a set of services. It's super for development.
+
+_Kubernetes_, however, is designed to help you run your application in a highly available mode using clusters and service replications. It is production grade.
+
+Describing Compose services as a Kubernetes application requires an extra layer that translates concepts from one to the other.
+
+Furthermore, on Kubernetes, you will share and promote your app through different stages. These stages require varying settings that should ideally be captured from the start.
```
[the remark this answers is not in the document]
Comment (ezodude) [reply README.md:211]: Thx! the above explains things more clearly.

Just rephrased slightly to work in this context:
```
Furthermore, on Kubernetes, you might also want to deploy or promote your app to different "stages", commonly known as environments. Application configuration may vary depending on the environment it is deployed to due to various infrastructure or operational constraints.

So, a good approach to managing your app configuration in different environments is a must.
```

File: README.md [hunk 25]
```diff
@@ -31,72 +42,120 @@ _Kev_ reduces the need for Kubernetes expertise in the team. The generated Kuber
...
 
-### Initiate a project
+## Quickstart
 
-Run the following command from within your project directory to initialise Kev project:
+- `kev init` - identifies a project's Compose Kubernetes source files and creates Compose environment overrides.
+- `kev render` - detects, applies any config changes and generates deployment manifests.
+- `kev help` - run it if you're a little lost.
+
+### init
```
Comment (marcinc) [inline README.md:55]: tiny one: make `init` and `render` (below) bold so they stand out more.

**Comment by marcinc** [comment]
Looks good, one final comment. Let's move `Contents` section below Features?

File: README.md (hunk 25 above)
Comment (ezodude) [reply README.md]: Reverted to older suggestions. They actually work better.

**Comment by ezodude** [comment]
- Moved TOC under features.
- Formatted TOC to make it stand out.
- Made the Quickstart more accessible in the TOC.
- Updated the Quickstart `init` and `render` headings as per older comments (they make more sense).
- Added a tutorial synopsis in the README to help readers understand what to expect.

**Review by marcinc — APPROVED** [review]
LGTM

**Pull request merged**